playground/src/auto-importer/auto-importer.js

```diff
@@ -40,28 +40,28 @@ class AutoImporter {
 
   constructor(componentsProvider, editor) {
     this.componentsProvider = componentsProvider;
-    this.componentsMap = {};
+    this.componentsMap = componentsProvider.get();
     this.editor = editor;
     Object.keys(ENGINE_MAP).forEach((k) => ENGINE_SET.add(ENGINE_MAP[k]));
   }
 
   update(validationResult) {
-    if (!Object.keys(this.componentsMap).length) {
-      this.componentsMap = this.componentsProvider.getComponents();
-    }
-    if (validationResult.status !== 'FAIL') {
-      return;
-    }
-    const missing = this._parseMissingElements(validationResult);
-
-    if (Object.keys(missing.missingTags).length
-        || missing.missingBaseScriptTag) {
-      const existing = this._parseHeadTag();
-      // The action taken to insert any elements to fix the report of missing
-      // tags is determined by both a combination of looking at the list of
-      // missing elements and also the current state of the <head> tag.
-      this._insertMissingElements(missing, existing);
-    }
+    this.componentsMap.then((components) => {
+      this.components = this.components || components;
```

Comment:
I'd prefer if you'd make the state handling more explicit by passing the components variable to _insertMissingElement instead of assigning it to a field.

Reply:
Done